interlok-core/src/main/java/com/adaptris/core/management/config/SharedConnectionConfigurationChecker.java

```diff
@@ -0,0 +1,112 @@
+package com.adaptris.core.management.config;
+
+import java.io.InputStream;
+import javax.xml.parsers.DocumentBuilder;
+import javax.xml.parsers.DocumentBuilderFactory;
+import javax.xml.xpath.XPath;
+import javax.xml.xpath.XPathConstants;
+import javax.xml.xpath.XPathExpressionException;
+import javax.xml.xpath.XPathFactory;
+import org.w3c.dom.Document;
+import org.w3c.dom.NodeList;
+import com.adaptris.core.management.BootstrapProperties;
+import com.adaptris.core.management.UnifiedBootstrap;
+import com.adaptris.core.util.DocumentBuilderFactoryBuilder;
+
+public class SharedConnectionConfigurationChecker implements ConfigurationChecker {
+  
+  private static final String FRIENDLY_NAME = "Shared connection check.";
+    
+  private static final String XPATH_AVAILABLE_CONNECTIONS = "//shared-components/connections/*/unique-id";
+  
+  private static final String XPATH_REFERENCED_CONSUME_CONNECTIONS = "//consume-connection/lookup-name";
+
+  private static final String XPATH_REFERENCED_PRODUCE_CONNECTIONS = "//produce-connection/lookup-name";
+  
+  private static final String XPATH_REFERENCED_SERVICES = "//connection[@class='shared-connection']/lookup-name";
+
+  @Override
+  public ConfigurationCheckReport performConfigCheck(BootstrapProperties bootProperties, UnifiedBootstrap bootstrap) {
+    ConfigurationCheckReport report = new ConfigurationCheckReport();
+    report.setCheckName(this.getFriendlyName());
+    
+    try {      
+      Document xmlDocument = buildXmlDocument(bootProperties);
+      
+      NodeList availableConnections = scanForConnections(xmlDocument, XPATH_AVAILABLE_CONNECTIONS);      
+      NodeList referencedConsumeConnections = scanForConnections(xmlDocument, XPATH_REFERENCED_CONSUME_CONNECTIONS);
+      NodeList referencedProduceConnections = scanForConnections(xmlDocument, XPATH_REFERENCED_PRODUCE_CONNECTIONS);
+      NodeList referencedServiceConnections = scanForConnections(xmlDocument, XPATH_REFERENCED_SERVICES);
+            
+      // **********************************
+      // Check all shared connections are used.
+      for(int counter = 0; counter < availableConnections.getLength(); counter ++) {
```

Comment:
```
      for(int counter = 0; counter < availableConnections.getLength(); counter ++) {
        if(!this.testReferenceExistsInNodeSets(availableConnections.item(counter).getTextContent(), referencedConsumeConnections)
            && !this.testReferenceExistsInNodeSets(availableConnections.item(counter).getTextContent(), referencedProduceConnections)
            && !this.testReferenceExistsInNodeSets(availableConnections.item(counter).getTextContent(),
                referencedServiceConnections)) {
          report.getWarnings().add("Shared connection unused: " + availableConnections.item(counter).getTextContent());
        }
      }
```

Is actually what you want in the check, since you're only checking that the available connections are being used in the consume connections...

but then I got to thinking; why aren't just using `com.adaptris.util.text.xml.XPath#selectMultipleTextItems()` instead, to simplify matters, since you're using a NodeList everywhere but you probably *don't need to*

Reply:
Also, using `c.a.u.t.x.XPath` means you can actually use Saxon+XPath2 features, since nowadays Saxon doesn't register as an XPathFactory so XpathFactory.newInstance() only ever gives you the com.sun.xerces one.

Reply:
Not checked at all (might not even compile!). It would have the added benefit on reporting on _all the produce connections that refer to a non-existent shared-connection_ ?

```
import com.adaptris.util.text.xml.XPath;
  public ConfigurationCheckReport performConfigCheck(BootstrapProperties bootProperties, UnifiedBootstrap bootstrap) {
    ConfigurationCheckReport report = new ConfigurationCheckReport();
    report.setCheckName(this.getFriendlyName());
    
    try {      
      Document xmlDocument = buildXmlDocument(bootProperties);
      XPath xpath = new XPath();

      List<String> availableConnections = Arrays.asList(xpath.selectMultipleTextItems(xmlDocument, XPATH_AVAILABLE_CONNECTIONS));
      List<String> referencedConsumeConnections = Arrays.asList(xpath.selectMultipleTextItems(xmlDocument, XPATH_REFERENCED_CONSUME_CONNECTIONS));
      List<String> referencedProduceConnections = Arrays.asList(xpath.selectMultipleTextItems(xmlDocument, XPATH_REFERENCED_PRODUCE_CONNECTIONS));
      List<String> referencedServiceConnections = Arrays.asList(xpath.selectMultipleTextItems(xmlDocument, XPATH_REFERENCED_SERVICES));

      Collection<String> union =
          CollectionUtils.union(CollectionUtils.union(referencedConsumeConnections, referencedConsumeConnections),
          referencedServiceConnections);

      // **********************************
      // Check all shared connections are used.
      for (String shared : availableConnections) {
        if (!union.contains(shared)) {
          report.getWarnings().add("Shared connection unused: " + shared);
        }
      }

      // Since we are going to discard them now, doing a remove while ill-advised is probably ok.
      referencedConsumeConnections.removeAll(availableConnections);
      if (referencedConsumeConnections.size() > 0) {
        report.setFailureException(new ConfigurationException(
            "Consume shared connection does not exist in shared connections: " + referencedConsumeConnections));
      }

      referencedProduceConnections.removeAll(availableConnections);
      if (referencedProduceConnections.size() > 0) {
        report.setFailureException(new ConfigurationException(
            "Consume shared connection does not exist in shared connections: " + referencedProduceConnections));
      }

      referencedServiceConnections.removeAll(availableConnections);
      if (referencedServiceConnections.size() > 0) {
        report.setFailureException(new ConfigurationException(
            "Service shared connection does not exist in shared connections: " + referencedServiceConnections));
      }
      
    } catch (Exception ex) {
      report.setFailureException(ex);
    }
    
    return report;
  }
```